torch/fx/passes/utils/matcher_utils.py

```diff
@@ -0,0 +1,230 @@
+from dataclasses import dataclass, field
+from collections import defaultdict
+import copy
+from torch.fx.graph import Graph
+from torch.fx.node import Node
+from torch.fx._compatibility import compatibility
+from typing import Dict, List, Set
+
+__all__ = ['SubgraphMatcher']
+
+
+@compatibility(is_backward_compatible=False)
+@dataclass
+class _InternalMatch():
+    # Node from which the match was found
+    anchor: Node
+    # Maps nodes in the pattern subgraph to nodes in the larger graph
+    nodes_map: Dict[Node, Node] = field(default_factory=dict)
+
+    # nodes in target graph that are matched placeholder in pattern
+    placeholder_nodes: List[Node] = field(default_factory=list)
+
+    # sink nodes in matched subgraph returned by output
+    returning_nodes: List[Node] = field(default_factory=list)
+
+    def __copy__(self):
+        return _InternalMatch(anchor=self.anchor, nodes_map=self.nodes_map.copy(),
+                              placeholder_nodes=self.placeholder_nodes.copy(),
+                              returning_nodes=self.returning_nodes.copy())
+
+@compatibility(is_backward_compatible=False)
+class SubgraphMatcher:
+    def __init__(self, pattern: Graph,
+                 match_output: bool = False,
+                 match_placeholder: bool = False,
+                 fully_contained: bool = True,
+                 remove_overlapping_matches: bool = True) -> None:
+        """
+        Args:
+            pattern: the targeted matching pattern, represented in fx.Graph
+            match_output: If True, output node in the pattern graph will be treated as a part of the targeted pattern.
+                If False, output node is ignored during match.
+            match_placeholder: If True, placeholder node in the pattern graph will be treated as a part of
+                the targeted pattern. If False, placeholder nodes will be used a wildcard.
+            fully_contained: If True, nodes (except placeholder and output_link_nodes ) can only be consumed
+                by nodes within th pattern.
+            remove_overlapping_matches: If True, in the case of overlapping matches, only the first match
+                will be returned.
+        """
+
+        self.pattern = pattern
+        self.match_output = match_output
+        self.match_placeholder = match_placeholder
+        self.fully_contained = fully_contained
+        self.remove_overlapping_matches = remove_overlapping_matches
+
+        if len(pattern.nodes) == 0:
+            raise ValueError("SubgraphMatcher cannot be initialized with an empty pattern")
+
+        for node in pattern.nodes:
+            if node.op != "output":
+                assert len(node.users) > 0, \
+                       "SubgraphMatcher cannot be initialized with an pattern with dead code"
+
+        # TODO: assert pattern is a connected graph
```

Comment:
Consider:

```
x1 = f1(a)
x2 = f2(a)
return x1, x2
```

sometimes it is profitable to fuse f1 and f2 together, as they share an input node. Is this expressible?

Reply:
yes, I added a test case for this pattern.